@Ch3LL here you go, I've targetted it against the `freeze` branch as I've noticed the issue is in Phosphorus milestone.


I've tested following usecases manually:

1. Argument passthrough
```bash
root@ip-172-31-45-201:~# cat test_args.py 
import sys

print(sys.argv)
root@ip-172-31-45-201:~# /tmp/tmpm5yodb3j/salt/dist/salt python test_args.py arg1 arg2 arg3
['test_args.py', 'arg1', 'arg2', 'arg3']
```

2. Exit code passthrough
```bash
root@ip-172-31-45-201:~# cat test_exit.py 
import sys

sys.exit(42)
root@ip-172-31-45-201:~# /tmp/tmpm5yodb3j/salt/dist/salt python test_exit.py 
root@ip-172-31-45-201:~# echo $?
42
```

3. Salt library interaction (the freeze branch still seems to represent as 2016.3.1 version)
```bash
root@ip-172-31-45-201:~# cat test_salt_shebang.py 
#!/tmp/tmpm5yodb3j/salt/dist/salt python
import salt.version

print("\n".join(salt.version.versions_report()))
root@ip-172-31-45-201:~# /tmp/tmpm5yodb3j/salt/dist/salt python test_salt_shebang.py 
Salt Version:
          Salt: 2016.3.1+46009.gbe2c8d402b
 
Dependency Versions:
          cffi: 1.14.6
      cherrypy: 18.6.1
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: 4.0.9
     gitpython: 3.1.27
        Jinja2: 3.1.0
       libgit2: Not Installed
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 1.0.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.17
      pycrypto: Not Installed
  pycryptodome: 3.9.8
        pygit2: Not Installed
        Python: 3.10.4 (main, Jul 19 2022, 11:24:48) [GCC 11.2.0]
  python-gnupg: 0.4.8
        PyYAML: 5.4.1
         PyZMQ: Not Installed
         smmap: 5.0.0
       timelib: 0.2.4
       Tornado: 4.5.3
           ZMQ: Not Installed
 
System Versions:
          dist: ubuntu 22.04 jammy
        locale: utf-8
       machine: x86_64
       release: 5.15.0-1011-aws
        system: Linux
       version: Ubuntu 22.04 jammy
```

4. Shebang functionality
```bash
root@ip-172-31-45-201:~# cat test_salt_shebang.py 
#!/tmp/tmpm5yodb3j/salt/dist/salt python
import salt.version

print("\n".join(salt.version.versions_report()))
root@ip-172-31-45-201:~# ./test_salt_shebang.py 
Salt Version:
          Salt: 2016.3.1+46009.gbe2c8d402b
 
Dependency Versions:
          cffi: 1.14.6
      cherrypy: 18.6.1
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: 4.0.9
     gitpython: 3.1.27
        Jinja2: 3.1.0
       libgit2: Not Installed
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 1.0.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.17
      pycrypto: Not Installed
  pycryptodome: 3.9.8
        pygit2: Not Installed
        Python: 3.10.4 (main, Jul 19 2022, 11:24:48) [GCC 11.2.0]
  python-gnupg: 0.4.8
        PyYAML: 5.4.1
         PyZMQ: Not Installed
         smmap: 5.0.0
       timelib: 0.2.4
       Tornado: 4.5.3
           ZMQ: Not Installed
 
System Versions:
          dist: ubuntu 22.04 jammy
        locale: utf-8
       machine: x86_64
       release: 5.15.0-1011-aws
        system: Linux
       version: Ubuntu 22.04 jammy

```

5. Relative imports
```bash
root@ip-172-31-45-201:~# cat test_app/__init__.py 
import sub

print(sub.test())
root@ip-172-31-45-201:~# cat test_app/sub.py 
def test():
    return "ok"
root@ip-172-31-45-201:~# /tmp/tmpm5yodb3j/salt/dist/salt python test_app/__init__.py 
ok
```

6. Exception handling
```bash
root@ip-172-31-45-201:~# cat test_exception.py 
raise Exception("test")
root@ip-172-31-45-201:~# /tmp/tmpm5yodb3j/salt/dist/salt python test_exception.py 
Traceback (most recent call last):
  File "/root/.pyenv/versions/3.10.4/envs/tmpyprx9b6s/bin/salt", line 113, in <module>
    redirect(sys.argv)
  File "/root/.pyenv/versions/3.10.4/envs/tmpyprx9b6s/bin/salt", line 93, in redirect
    python_runtime()
  File "/root/.pyenv/versions/3.10.4/envs/tmpyprx9b6s/bin/salt", line 70, in python_runtime
    exec(f.read())
  File "<string>", line 1, in <module>
Exception: test
[85680] Failed to execute script 'salt' due to unhandled exception!
[ERROR   ] An un-handled exception was caught by Salt's global exception handler:
Exception: test
Traceback (most recent call last):
  File "/root/.pyenv/versions/3.10.4/envs/tmpyprx9b6s/bin/salt", line 113, in <module>
    redirect(sys.argv)
  File "/root/.pyenv/versions/3.10.4/envs/tmpyprx9b6s/bin/salt", line 93, in redirect
    python_runtime()
  File "/root/.pyenv/versions/3.10.4/envs/tmpyprx9b6s/bin/salt", line 70, in python_runtime
    exec(f.read())
  File "<string>", line 1, in <module>
Exception: test
```

Here I'm a bit unsure whether I shouldn't rather wrap the whole `exec` in try/except and just print the exception and exit with 1 (as Python does it), because I don't particularly like the Salt exception handler mixing things up here



And also the tests are not written in Salt test suite as I don't believe there is any example test I could use to test this behaviour. Will ask on Slack if anybody has any idea.

--- a/run.py
+++ b/run.py
@@ -65,9 +65,13 @@ def python_runtime():
 
     # update passed args so they don't start with "<binary> python"
     sys.argv = sys.argv[2:]
-
-    with open(script) as f:
-        exec(f.read())
+    exec_locals = {"__name__": "__main__", "__file__": str(script), "__doc__": None}
+    with open(script, encoding="utf-8") as rfh:
+        try:
+            exec(rfh.read(), exec_locals)
+        except Exception:
+            traceback.print_exc()
+            sys.exit(1)
 
 
 def redirect(argv):

--- a/run.py
+++ b/run.py
@@ -61,7 +61,6 @@ def python_runtime():
     # extract the absolute script path to alter sys.path and specific dunder variables
     script = pathlib.Path(sys.argv[2]).expanduser().absolute()
     sys.path.insert(0, str(script.parent))
-    __file__ = str(script)
 
     # update passed args so they don't start with "<binary> python"
     sys.argv = sys.argv[2:]

--- a/run.py
+++ b/run.py
@@ -59,11 +59,11 @@ def py_shell():
 
 def python_runtime():
     # extract the absolute script path to alter sys.path and specific dunder variables
-    script = pathlib.Path(sys.argv[2]).expanduser().absolute()
+    script = pathlib.Path(sys.argv[2]).expanduser().resolve()
     sys.path.insert(0, str(script.parent))
 
     # update passed args so they don't start with "<binary> python"
-    sys.argv = sys.argv[2:]
+    sys.argv[:] = sys.argv[2:]
     exec_locals = {"__name__": "__main__", "__file__": str(script), "__doc__": None}
     with open(script, encoding="utf-8") as rfh:
         try:

